redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt

```diff
@@ -127,242 +128,187 @@ internal class UIViewBox :
-          CrossAxisAlignment.Center -> ((frameHeight - viewHeight - margin.height) / 2.0) + margin.top.toPx()
-          CrossAxisAlignment.End -> frameHeight - viewHeight - margin.bottom.toPx()
-          else -> margin.top.toPx()
-        }
-
-        // Position the view.
-        view.setFrame(CGRectMake(x, y, viewWidth, viewHeight))
-      }
     }
 
-    override fun sizeThatFits(size: CValue<CGSize>): CValue<CGSize> {
-      val zero = CGSizeMake(0.0, 0.0)
-
-      val requestedSize = children.widgets.fold(zero) { acc, widget ->
-        maxEachDimension(acc, widget.modifier.requestedSize)
-      }
-
-      val wrapOrFillSize = when {
-        widthConstraint == Constraint.Wrap || heightConstraint == Constraint.Wrap -> {
-          val wrapSize = children.widgets.fold(zero) { acc, widget ->
-            maxEachDimension(acc, widget.value.sizeThatFits(size) + widget.modifier.margin)
-          }
-
-          CGSizeMake(
-            widthSize = when (widthConstraint) {
-              Constraint.Wrap -> wrapSize
-              else -> size
-            },
-            heightSize = when (heightConstraint) {
-              Constraint.Wrap -> wrapSize
-              else -> size
-            },
-          )
-        }
+    val availableWidth = (frameWidth - marginWidth).coerceAtLeast(0.0)
+    val availableHeight = (frameHeight - marginHeight).coerceAtLeast(0.0)
 
-        // Optimization: Don't call sizeThatFits() if we don't need to.
-        else -> size
-      }
-
-      return maxEachDimension(requestedSize, wrapOrFillSize)
+    val fitWidth = when {
+      !requestedWidth.isNaN() -> requestedWidth
+      horizontalAlignment == CrossAxisAlignment.Stretch -> availableWidth
+      else -> availableWidth
+    }
+    val fitHeight = when {
+      !requestedHeight.isNaN() -> requestedHeight
+      verticalAlignment == CrossAxisAlignment.Stretch -> availableHeight
+      else -> availableHeight
     }
-  }
-}
 
-/**
- * Returns the literal size specified by the width, height, size, and margin modifiers.
- *
- * If no size is specified this returns `0.0 x 0.0`.
- *
- * If conflicting sizes are specified this returns the maximum of the values.
- */
-internal val Modifier.requestedSize: CValue<CGSize>
-  get() {
-    var width = 0.0
-    var height = 0.0
-    var margin = Margin.Zero
-
-    forEachScoped { childModifier ->
-      when (childModifier) {
-        is Width -> with(Density.Default) {
-          width = maxOf(width, childModifier.width.toPx())
-        }
+    // Measure the view if don't have an exact width or height.
+    val mustMeasureWidth = requestedWidth.isNaN() &&
+      horizontalAlignment != CrossAxisAlignment.Stretch
+    val mustMeasureHeight = requestedHeight.isNaN() &&
+      verticalAlignment != CrossAxisAlignment.Stretch
 
-        is Height -> with(Density.Default) {
-          height = maxOf(height, childModifier.height.toPx())
-        }
+    if (!mustMeasureWidth && !mustMeasureHeight) {
```

Comment:
Android doesn’t have this optimization. It might benefit from it.

Reply:
(It’s more difficult because on Android measuring may have essential side-effects)